### ISSUE TITLE
[Bug] Coroutines, Listening for the OAC, Blocking Sockets Dilemmas in Refactoring

### ISSUE THREAD

**Issue by Spydrouge** [opencog/unity3d-opencog-game#49]
### Synopsis

Included is a journal of an exploration into solving certain bugs. Many of these were later solved, but the biggest one that was never truly "solved" is that if one calls /load AGI_ROBOT while the oac is initializing Linux side, the game may crash in catastrophic ways- and it had everything to do with threads, coroutines, and the fact that we just didn't have time to go research new and improved Unity-kosher ways of doing things and implement them.

My journal is left here as documentation, because it took me forever to figure out that a Write command on a stream on the main thread would, if the game was started just wrong in a narrow window, accidentally block up a socket and cause the game to hang. I worked around this by catching a message from the oac that it was unavailable which would then shut the game down prematurely (before the hang could ever happen) but I'm sure you could still replicate the hang if you really tried.

To fix the hang, one would have to think about how one wanted to implement it (with threads, with rewriting the code to use unity networking API, etc) and then execute one's plan. And since I got around all the bugs, that was code we just didn't need at the moment.

But because I found out a lot of things on the way, I've left this here as documentation for le future.
### Tracking Down the Problem (Journalling)

Right now sensing the network for the unit tests is going very smoothly (in fact writing the unit tests has helped me fix a few booleans, flags, and others checks so that messages don't get sent to the OAC too fast and so that we don't have to WaitForSeconds(1f) everywhere in big loops to try and slow down for it. (I think right now we only have a single race-preventer, and it's filed under a FIXME.

Anyhoodie, we still have a fun problem.

If one starts up the game engine too early while the embodiment is trying to initialize, the sockets will all connect and the OAC will even send back a message to say it is initialized (or try, anyway.) which we are now correctly sensing for.

But the OAC will end up imploding and will mark itself as unavailable and never recover.

Now there is a way to catch this by saying 'if ever the OAC marks itself in this condition, do X' but I also want to look at why the Unit Tests never fail. Necessarily it must be the case that the FIRST unit test (which never passes or fails) ends up calling a function that exists in an infinite loop. 

So what we should do is track down culprits. 

We get up to the printing of this statement:
`UnityEngine.Debug.Log(OCLogSymbol.CONNECTION + "Socket accepted...");`

Which occurs after
`Debug.Log(OCLogSymbol.CONNECTION + "Waiting for OAC...");`

The only explanation for the fact that we never Make it to `LoadMethod's` debug statements is that we get stuck somewhere in `LoadOAC();`

This function is not very long:

```
System.Text.StringBuilder msgCmd = new System.Text.StringBuilder("LOAD_AGENT ");
msgCmd.Append(_brainID + WHITESPACE + _masterID + WHITESPACE);
msgCmd.Append(_type + WHITESPACE + _traits + "\n");

OCMessage msg = OCMessage.CreateMessage(_ID,
                      OCConfig.Instance.get("SPAWNER_ID"),
                      OCMessage.MessageType.STRING,
                      msgCmd.ToString());
SendMessage(msg);
```

CreateMessage is essentially nothing more than a constructor. That leaves us only `SendMessage` to be stuck in.  And in send message, the only place to go where we might be getting stuck is in this command: `bool result = Send(command.ToString());`

Unfortunately we arrive to find little useful. There are no loops. Nothing telescopes further that we can see. All code paths return a boolean variable. Where could we be getting stuck? Back in `Send Message` we have code under _sockets.Count == 1 that yields nothing, but the message above it isn't displayed in multiplicity. 

There is a loop in Listen(), where we see `while(!_shouldStop){}` Of the two control paths, the first would be yielding back and allowing us to catch the timeout. The second would be repeatedly printing debug messages. 

But hey; how did we ever get to Listen()? I don't see a path from LoadOAC to here. Listen() is only ever began by StartCoroutine(), through StartListening();

StartListening(); is only ever called by public void Update(); So it could be that Listen() was activated in parallel  to the unit test coroutines. But I must remember this confusion point: Coroutines don't actually operate in different threads. YES OldMessagEHandler does have a separate thread going on, and yes startup and update occur on slightly different threads... But Update() shouldn't be getting called midway through one of my Unit Test coroutines entering an infinite loop.

Perhaps it is as simple as me getting some notation wrong somewhere.  Here is the code we are trying to get working

```
yield return StartCoroutine(GameManager.entity.load.AtRunTime(embodimentSpawn, embodimentPrefab, "", masterName, masterId, report));

if(didConnect)
{
    Debug.Log(OCLogSymbol.PASS + "Testing the Embodiment, Succeeded");
}
else
{
    Debug.Log(OCLogSymbol.FAIL + "Testing the Embodiment, Failed");
    results[(int)TestTypes.EMBODIMENT] = false;
    result = false;
}
yield break;
```

In AtRunTime we do this:

```
//and try to connect
yield return StartCoroutine(connector.ConnectOAC());

//if we failed to initialize the connector
if (!connector.IsInitialized) 
{

    // OAC is not loaded normally, destroy the avatar instance.
    Debug.LogError (OCLogSymbol.ERROR + "Could not connect to Embodiment");
    System.Console.WriteLine(OCLogSymbol.DETAILEDINFO + "LoadMethods.AtRunTime is reporting !connector.Initialize. Cannot connect to the OAC, avatar loading failed.");
    connector.SaveAndExit ();
    Destroy (agentClone);

    if(report != null)
        report("false");
} 
else
{
    Debug.Log (OCLogSymbol.CLEARED + "LoadMethods.AtRunTime is reporting connector.Initialized");
    if(report != null)
        report("true");
}

yield break;
```

Again I am seriously lost as to how NOTHING is every yielding back to UnitTests. I'll try running with the debugger attached and pop a breakpoint halfway through. 

You can tell I'm lost because nowhere I popped a breakpoint actually fired XD. I tried hitting 'pause' in the hopes it might be able to bring me ANYWHERE interesting. Unfortunately it looks like it killed Mono instead. 

Ooh! It finally found where we are. According to it, we're in the OldMessageHandler thread, and we're looking at this: `public void run()`  Which has an infinite `while( !endInput )` loop. But nothing looks to be going on in there that would block the main thread from executing, and unfortunate it looks like Mono nabbed this side thread instead of the main thread...

Mono does have a Threads panel, but we only have a location on Thread 2, which appears to be the OldMessageHandler. 

(Note, I've determined that this: `console.AddConsoleEntry("First callback received, initializing MessageHandler...", "Unity World", OpenCog.Utility.Console.Console.ConsoleEntry.Type.RESULT);` prints. However, again, it is in Listen() which is part of an Update() function. 

I have noticed that thread 1 is the main thread and thread 2 is the OldMessageHandler

**Debug.Log around LoadOAC**
LoadOAC is never exiting, and I just cannot figure out why. 

**Something to Check**
Could the explanation for the never-finishing Coroutines be- at all- that the coroutines THEY were calling suddenly turned up null? Let's see. We know UnitTests and AtRunTime did not dissappear.-
### WAIT. It finally occurred to me

I was looking at those System.IO.Stream darlings for the longest time. I glanced at them and then glanced away going 'naww, can't be' Well I decided to do a google search on 'unity Stream freeze' and it turns out that stream readers and writiers are _blocking calls._ It is normal to put them on a separate thread because they will block WHATEVER thread they are on until data arrives or is received. 

Really.  This is why we have OldMessageHandler operating in another thread.

Now oridnarily this would only shoot us in the foot if we were READING from a message stream. Not writing. Writing we can just pop our message in and send it on it's way, right? Wrong. In the event of a faulty connection, a Write call on a stream will hang. So the Unit Tests will hang because the code to create the agent hangs, because the loadOAC hangs, because a call to sw.Write() hangs.

So let's pause and look at this for a second. What can we do differently? 

**Ignore it**
Create an issue tagged "WontFix" and notify everyone that it's something for future releases, and not to be dumb about when they start the game (It's a very small window after all)

**Hacky Solution**
OldMessageHandler is going to pick up the 'spawner is unavailable' message, which is a clear indicator at this point that everything went Kablooey. I'm not 100% sure it can rescue the main the thread (could it destroy the connector singleton? could it get UnitTests firing again by somehow seperating the coroutines? For example, is it possible to use the string based StartCoroutine somewhere and then call STopCoroutine to try and rescue UnitTests- or would that only stop it from ever going to the MoveNext() function and so basically leave everything in exactly the same state it started? At the very least it could produce an error message "Embodiment did not initialize before game started!" and crash unity. (I still don't know how the unit tests are going to report back failure, but I'm sure someone's got a plan for that.)

**Legit And Scary Solution**
Create an OldMessageSender XD. Run it on thread 3 or 4 or whatever. Copy just the Send() code into it. Things will then timeout naturally at least! 

**Legit and Even Scarier Solution**
Why are we using TCP sockets at all? Why not just use Unity network functions- the API that's built into the tool we're using? 

**Possibilities to Study: Flags**
Is there a flag somewhere we ought to be looking at? Is it possible we can put a test in Send() that would prevent it from launching the blocking code? AFter all, we manage not to freeze anywhere else, even when connections totally fail. 

**Possibilities to Study: _messagesToSend.Add()**
The recommendations online suggest coroutines. Initially I dismissed this because I thought "If it hangs, it's never going to get to the WaitForSeconds() yield! Then I realized they meant to start multiple co-routines from the Start(). Although I've been told these coroutines run linearly, the implication seems to be that, in the even that one hangs, it still won't derail the whole system (how is that possible? I don't know) 

Actually no, look at this:

> You cannot and should not.
> 
> Coroutines are a form of cooperative multithreading that are actually run on the main thread. Since you cannot interrupt NetworkStream.GetStream() it cannot be a coroutine. And you certainly don't want your network I/O to block the main thread.
> 
> Network I/O definitively belongs in a separate worker thread.

**Possibilities to Study: Does anywhere else in my code handle this?**
That's hard to say without quite a bit of study! But While I was seeing many different answers (most of which said 'build another thread!' I did come across this: 

> To prevent your tcpListener to block until a connection comes in, you can use TCPListener.pending() to check if a connection is pending and just then call "AcceptTCPClient".

I have noticed that in other areas where we are sending messages, we DO ask whether or not the connection is pending. Yes! Under OCServerListener, we have that Listen() command. Obviously that's for a listener, but maybe we can -

**Possibilities to Study: What is a nonblocking socket?**

There is an accessor under _clientSocket called Blocking. It can be set.  :/
Hmm. I tried setting it, and now we get an error in our try/catch: Operation not allowed on a non-blocking socket.  According to the internetz this is because our send buffer fills up and we can't send the next message. My question is, can we encapsulate this in a loop with a yield? 

How does one use non-blocking sockets? I found some code samples for non-blocking sockets [here](http://www.kegel.com/dkftpbench/nonblocking.html).

**Possibilities to Study: What is an async socket?**

> stream.Read is a blocking call, it will halt the entire thread where it's called until data arrives. This is the expected behavior, look into using either threads or async calls (begin/end pairs).

I'm seeing words like 'asyncSocket' which seems to suggest I am doing the thing he suggests to do. I wish I knew more about networking so I could come up with an idea of how this works. Maybe I'll ask Lake? 

Answering this by myself- using things like BeginWrite, etc- would take A LOT OF TIME. I need to remember how little time I have before I leave on Tuesday. This just can't be fixed by me right now. 

**Note: Reasons we need to be on the same network**

```
Socket asyncSocket = new 
                Socket
                (AddressFamily.InterNetwork
                , SocketType.Stream
                , ProtocolType.Tcp
                );
System.IAsyncResult ar = asyncSocket
            .   BeginConnect
                (ipe
                , new System.AsyncCallback(ConnectCallback)
                , asyncSocket
                );

_clientSocket = (Socket)ar.AsyncState;

```

**Note: Is the reason we can't run in webplayer?**

> Its also a known problem here in the community that any use of async callback can and will kill the webplayer on shutdown and with it the browser process (you will often also see a significantly raised unity editor crash probability as the editor behaves like the webplayer on game end ...). for that reason, working with chrome during development can save quite some time, as only a tab will die when this happens not the whole browser.
